File: scala/bob/src/main/scala/Bob.scala
object Bob {
  def response(statement: String): String = {
    statement.trim match {
      case s if s.isEmpty =>
        "Fine. Be that way!"
      case s if isShoutingQuestion(s) =>
        "Calm down, I know what I'm doing!"
      case s if isQuestion(s) =>
        "Sure."
      case s if isShouting(s) =>
        "Whoa, chill out!"
      case _ =>
        "Whatever."
    }
  }

  def isShoutingQuestion(statement: String): Boolean =
    statement.toUpperCase == statement &&
      hasLetters(statement) &&
      isQuestion(statement)

  def isShouting(statement: String): Boolean =
    statement.toUpperCase == statement &&
      hasLetters(statement)

  def isQuestion(statement: String): Boolean =
    statement.endsWith("?")

  def hasLetters(statement: String): Boolean =
    statement.toLowerCase.exists(('a' to 'z').contains(_))
}

File: clojure/bob/src/bob.clj
(ns bob
  (:require [clojure.string :as string]))

(defn yelling?
  "Checks if shouting if all capitals and has letters."
  [s]
  (and (some #(Character/isLetter %) s)
       (= (string/upper-case s) s)))


(defn response-for [s]
  (let [s (string/trim s)]
    (cond
      (and (yelling? s)
           (string/ends-with? s "?")) "Calm down, I know what I'm doing!"
      (string/ends-with? s "?") "Sure."
      (string/blank? s) "Fine. Be that way!"
      (yelling? s) "Whoa, chill out!"
      :else "Whatever.")))

